examples/vision/nerf.py

```diff
@@ -0,0 +1,676 @@
+    t_vals = tf.linspace(near, far, num_samples)
+    if rand:
+        # Inject uniform noise into the sample space to make the sampling
+        # continuous.
+        shape = list(ray_oris.shape[:-1]) + [num_samples]
+        noise = tf.random.uniform(shape=shape) * (far - near) / num_samples
+        t_vals = t_vals + noise
+
+    # r(t) = o + td -> Building the "r" here.
+    rays = ray_oris[..., None, :] + (ray_dirs[..., None, :] * t_vals[..., None])
+    rays_flat = tf.reshape(rays, [-1, 3])
+
+    # Apply positional encdoing.
+    rays_flat = pos_encode(rays_flat)
+
+    return (rays_flat, t_vals)
+
+
+def map_fn(pose):
+    # Get the ray origin and directions.
+    (ray_oris, ray_dirs) = get_rays(height=H, width=W, focal=focal, pose=pose)
+
+    # Render the rays from the origin and direction.
+    (rays_flat, t_vals) = render_flat_rays(
+        ray_oris=ray_oris,
+        ray_dirs=ray_dirs,
+        near=2.0,
+        far=6.0,
+        num_samples=NUM_SAMPLES,
+        rand=True,
+    )
+    return (rays_flat, t_vals)
+
+    # get_rays
+    (ray_oris, ray_dirs) = get_rays(H, W, focal, pose)
+    rays_flat, t_vals = render_flat_rays(
+        ray_oris, ray_dirs, near=2.0, far=6.0, num_samples=NUM_SAMPLES, rand=False
+    )
+    return rays_flat, t_vals
+
+
+# Make the `tf.data` pipeline.
+train_img_ds = tf.data.Dataset.from_tensor_slices(images)
+train_pose_ds = tf.data.Dataset.from_tensor_slices(poses)
+
+# Map the pose ds to get "flat_rays" and "t_vals" as dataset.
+train_ray_ds = train_pose_ds.map(map_fn, num_parallel_calls=AUTO)
+
+# Zip the image and rays dataset together.
+training_ds = tf.data.Dataset.zip((train_img_ds, train_ray_ds))
+
+# Shuffle and batch the dataset.
+train_ds = (
+    training_ds.shuffle(BATCH_SIZE)
+    .batch(BATCH_SIZE, drop_remainder=True, num_parallel_calls=AUTO)
+    .prefetch(AUTO)
+)
+
+"""
+# NeRF model
+
+The model is a multi-layer perceptron. It has relu as its non-linearity. 
+
+An exerpt from the paper:
+
+"We encourage the representation to be multiview consistent by
+restricting the network to predict the volume density sigma as a
+function of only the location x, while allowing the RGB color c to be
+predicted as a function of both location and viewing direction. To
+accomplish this, the MLP first processes the input 3D coordinate x
+with 8 fully-connected layers (using ReLU activations and 256 channels
+per layer), and outputs sigma and a 256-dimensional feature vector.
+This feature vector is then concatenated with the camera ray’s viewing
+direction and passed to one additional fully-connected layer (using a
+ReLU activation and 128 channels) that output the view-dependent RGB
+color."
+
+Here we have gone for a minimal implementation and have used `64`
+Dense units instead of `256` as that mentioned in the paper.
+"""
+
+
+def get_nerf_model(num_layers, num_pos):
+    # Initialize the input layer.
+    inputs = tf.keras.Input(shape=(num_pos, 2 * 3 * POS_ENCODE_DIMS + 3))
```

Comment:
Your lines will be marginally shorter if you import `keras` and the `layers` module, then access members of these modules instead of `tf`